preprocessors/object-detection-azure/objdetect.py

```diff
@@ -0,0 +1,190 @@
+# GNU Affero General Public License for more details.
+# You should have received a copy of the GNU Affero General Public License
+# and our Additional Terms along with this program.
+# If not, see
+# <https://github.com/Shared-Reality-Lab/IMAGE-server/LICENSE>.
+
+import requests  # pip3 install requests
+from re import search
+import operator
+
+# import numpy as np
+import json
+import time
+import jsonschema
+import logging
+import base64
+import os
+from flask import Flask, request, jsonify
+
+app = Flask(__name__)
+
+
+def process_results(result):
+    send = []
+    image_height = result["metadata"]["height"]
+    image_width = result["metadata"]["width"]
+    for item in range(len(result['objects'])):
+        dims = result["objects"][item]['rectangle']
+        x_left = dims['x'] / image_width
+        x_right = (dims['x'] + dims['w']) / image_width
+        y_left = dims['y'] / image_height
+        y_right = (dims['y'] + dims['h']) / image_height
+        centre = [abs((x_left + x_right) / 2),
+                  abs((y_left + y_right) / 2)]
+        area = abs((dims['w'] * dims['h']) / (image_height * image_width))
+        class_prob = float(result['objects'][item]['confidence'])
+        normalised_dims = [x_left, y_left, x_right, y_right]
+        dictionary = {"ID": item,
+                      "type": result['objects'][item]['object'],
+                      "dimensions": normalised_dims,
+                      "confidence": class_prob,
+                      "centroid": centre,
+                      "area": area
+                      }
+        send.append(dictionary)
+    return send
+
+
+def process_image(image):
+
+    region = "canadacentral"  # For example, "westus"
+    api_key = os.environ["AZURE_API_KEY"]
+
+    # Set request headers
+    headers = dict()
+    headers['Ocp-Apim-Subscription-Key'] = api_key
+    headers['Content-Type'] = 'application/octet-stream'
+
+    # Set request querystring parameters
+    """Only query categories"""
+    params = {'visualFeatures': 'Objects'}
+
+    # Make request and process response
+    response = requests.request(
+        'post',
+        "https://{}.api.cognitive.microsoft.com/vision/v1.0/analyze".format(region),
+        data=image,
+        headers=headers,
+        params=params
+    )
+    if response.status_code == 200 or response.status_code == 201:
+
+        if 'content-length' in response.headers and \
+                int(response.headers['content-length']) == 0:
+            return "", 204
+        elif 'content-type' in response.headers and \
+                isinstance(response.headers['content-type'], str):
+            if 'application/json' in response.headers['content-type'].lower():
+                if response.content:
+                    result = response.json()
+                    label = process_results(result)
+                else:
+                    return "", 204
+            else:
+                return "", 204
+
+    else:
```

Comment:
If we got a not ok response from azure, this really needs to be logged at a warning/error level. As of now, it could fail silently.

Reply:
Note that this applies for all of the `return "", 204` lines in this part of the file.

Reply:
Is there a reference for the taxonomy of error codes ?

Reply:
Still unclear, more details needed.

Reply:
When our quota was reached, the request to Azure stopped returning data. If Azure becomes unreachable for a time, we don't get a 200 (OK) response. In these cases, what would happen here is the preprocessor would report nothing in logs, return 204 (No Content, what we use to say everything is okay and we are returning no data) and exit. In these cases, something has went wrong and if nothing is done to report it here we won't be able to diagnose the problem.